Add filters to Browse XComs

RoyLee1224 opened the pull request:
## Related Issue

closes: #53056

## Changes Made

Added search filters for `Key`, `Dag display name`, `Task ID`,`Run ID`, `logical date`, `run after` .
~~- **Note:** The current implementation requires an exact match for the filter values.~~

## Screenshot

[<img width="3024" height="1100" alt="basic_filters" src="https://github.com/user-attachments/assets/9621e9ba-7d62-41c6-9bb7-52930e9e2f65" />
](https://github.com/user-attachments/assets/a36efcd9-60d5-43f5-9dad-c5d665ec9d59)

## TO-DO

- [x] Add collapsible filter panel(Advanced filters) for better UX
- [x] Propose backend API extensions for pattern matching
- [x] Implement remaining filters (`Logical Date Range`, `Run After Range`)
- [x] Implement `map_index` filter with NumberInput
- [x] Hide some filters in the TI Xcom tab depending on the context

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

RoyLee1224 committed:
feat(ui): Add filters for Key, DAG, Task, and Run ID to XCom page

RoyLee1224 committed:
feat: implement basic search with filter pill

RoyLee1224 committed:
chore(ui): modify filterpill appearance

RoyLee1224 committed:
style(ui): Polish appearance of filter components

pierrejeambrun reviewed (commented):
Nice.

Do not incorporate the collapsing of filters etc... in this PR. (FilterManager.tsx) It is a separate one because we will probably need to discuss about the design. We should keep the same structure that other page filters, for instance in your example there is no 'reset filter' button, etc.

RoyLee1224 committed:
feat(api, ui): Add wildcard search to XComs and simplify filter UI

RoyLee1224 committed:
feat(api): Add dag, run, task wildcard search to XComs

fix(api): Refactor XCom entries filter

fix(api): Correct XCom endpoint filtering

RoyLee1224 committed:
refactor(ui): Simplify XCom filters

RoyLee1224 committed:
feat(api, ui): Add Logical Date Range, Run After Range filter to Xcoms

RoyLee1224 committed:
Merge branch 'main' into filters

RoyLee1224 committed:
fix(i18n): Modify translations

pierrejeambrun reviewed (commented):
> For the temporary UI in this PR, should I use the simple multi-input bar design in Airflow 2 as a baseline?


You can take a look at the `DagsList.tsx` page that implement a bunch of different filters. It should be similar in terms of look and feel.

RoyLee1224 committed:
fix: Update test to use xcom_key_pattern parameter

RoyLee1224 committed:
refactor(XCom): Unifying handlers and config

RoyLee1224 committed:
fix(XCom): Hide filters in task instance Xcom tab

RoyLee1224 committed:
feat(UI): Add reset button to XCom filters

Lee-W reviewed (approved):
backend and translation looks good to me

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -732,6 +733,20 @@ def _transform_ti_states(states: list[str] | None) -> list[TaskInstanceState | N
+]
+QueryXComTaskIdPatternSearch = Annotated[
+    _SearchParam, Depends(search_param_factory(XComModel.task_id, "task_id_pattern"))
+]
```
In the UI we most likely display `dag_display_name` but here we filter on `dag_id` which might confuse users.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/XCom/XCom.tsx:
```diff
@@ -103,25 +112,49 @@ export const XCom = () => {
       ) : undefined}
+
+      {dagId === "~" && runId === "~" && taskId === "~" ? <XComFilters /> : undefined}
+
```
It would be nice to still have other filtering option on details pages.

RoyLee1224 replied:
Good point. I hadn't considered mapped tasks scenario that make multiple results, I thought we would only get one result in that tab. 

I'll fix it.

pierrejeambrun commented on airflow-core/src/airflow/ui/public/i18n/locales/en/browse.json:
```diff
@@ -18,6 +18,16 @@
+      "runAfterTo": "Run After To",
+      "runIdFilter": "Filter by Run ID",
+      "taskIdFilter": "Filter by Task ID"
+    },
```
We most likely have some of those key already existing, we might want to regroup them. key with "Placeholder" suffix. Those should also have "Placeholder" suffix.

RoyLee1224 replied:
Sure, should I open another PR for regrouping placeholders in other files?

pierrejeambrun reviewed (commented):
We're probably missing the map_index filter too. It wasn't mention in the original issue, but I think we should add that too.

RoyLee1224 committed:
fix(ui, api): filter by dag display name & regroup placeholder translations

RoyLee1224 committed:
fix(ui): modify translations

RoyLee1224 committed:
feat(ui,api): Add mapIndex filter

RoyLee1224 committed:
fix(api,ui): Change map_index filter to exact match

RoyLee1224 committed:
fix(api,ui): Change map_index filter to exact match

jason810496 reviewed (commented):
Nice! Thanks for the PR!

Non-blocking:
It would be great to add some test cases for new filters in `airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_xcom.py`, thanks.

jscheffl reviewed (commented):
I also think this is looking good now!

One thing that jumps into my view now - but is something that might be a follow-up PR: The view in the tag of a task instance is fitlering on the task instance. But (1) filters as well as (2)  table columns are not needed for these pre-selections as dag, run, task, map-index are filtered. That might make the screen a bit cleaner inside the tab of the grid view:
<img width="1279" height="729" alt="image" src="https://github.com/user-attachments/assets/357ac1ed-8fba-4008-beba-c845e0288d84" />

could be
<img width="1265" height="684" alt="image" src="https://github.com/user-attachments/assets/0a2f5266-e5ed-467c-b099-c2a5f78237a6" />

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/xcom.py:
```diff
@@ -127,8 +139,16 @@ def get_xcom_entries(
     readable_xcom_filter: ReadableXComFilterDep,
     session: SessionDep,
-    xcom_key: Annotated[str | None, Query()] = None,
-    map_index: Annotated[int | None, Query(ge=-1)] = None,
```
We should keep both on the backend,`xcom_key` and `xcom_key_pattern`, that's a breaking change. Leave the exact search on `xcom_key` and do a search on `xcom_key_pattern`.

RoyLee1224 replied:
Sure, that makes perfect sense.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/xcom.py:
```diff
@@ -127,8 +139,16 @@ def get_xcom_entries(
+    task_id_pattern: QueryXComTaskIdPatternSearch,
+    map_index: Annotated[
+        FilterParam[int | None],
+        Depends(filter_param_factory(XComModel.map_index, int | None, filter_name="map_index")),
```
Nice.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/XCom/XCom.tsx:
```diff
@@ -103,25 +113,55 @@ export const XCom = () => {
+  };
+
+  const { data, error, isFetching, isLoading } = useXcomServiceGetXcomEntries(apiParams, undefined, {
+    enabled: !isNaN(pagination.pageSize),
```
```suggestion
```

Why is that necessary?

RoyLee1224 replied:
According to #44989 and its following PR, `enabled: !isNaN(pageSize)` check was a defensive fix for a race condition. 

And this issue seems to be resolved in #45095, which uses a router loader to ensure the config is always pre-loaded. 

I have locally removed this check and seems like the original bug does not reappear.

I found this same redundant check in several other files as well (e.g., `DagRuns.tsx`, `Events.tsx`,`TaskInstances.tsx`, `HITLTaskInstances.tsx`).

Should I remove this single instance here, or should I open a dedicated new PR to clean up all of them at once?
 I lean towards a separate PR for better tracking, but I'm happy to follow your guidance.

pierrejeambrun replied:
Yes you are right, a separate PR is better, lets keep following the pattern here then.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/XCom/XComFilters.tsx:
```diff
@@ -0,0 +1,180 @@
+    return (
+      <Box key={key} w="200px">
+        <Text fontSize="xs" marginBottom={1}>
+          {type === "search" ? "\u00A0" : translate(`common:filters.${translationKey}`)}
```
```suggestion
          {type === "search" ? "\u00A0" : translate(`common:filters.${translationKey}`)}
```

Can you explain `\u00A0`?

RoyLee1224 replied:
Using `\u00A0` ensures the search filter container has the same height as those time-related filters (e.g., "Logical Date from"). Without it, the search box would be higher than the rest. Like this:

<img width="2134" height="368" alt="filter" src="https://github.com/user-attachments/assets/fdb85622-e920-4e6b-8104-d743522c8e98" />

pierrejeambrun replied:
This can be achieved via CSS and we should prefer this because it's easier to understand.

RoyLee1224 replied:
Sure, I'll try.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/XCom/XComFilters.tsx:
```diff
@@ -0,0 +1,180 @@
+    return (
+      <Box key={key} w="200px">
+        <Text fontSize="xs" marginBottom={1}>
+          {type === "search" ? "\u00A0" : translate(`common:filters.${translationKey}`)}
```
It's hard to render filter in a generic way like this because we are most likely not specifying the proper input type.

For instance map_index should be a number, not a free `str` text.

pierrejeambrun replied:
@jscheffl is right, in the detail page we need to hide filters for parameters that are already set.

Basically on the global table `dag_id`, `run_id` `task_id` === `~` you can display additional search box for dag_display name, run_id, and task_id pattern.

On the detail of DagRun, Dag or TaskIntance, we need to hide the proper input.

For `task_id` and `dag_run_id` if those are set as path parameter (not query parameter and equal to `~`), then we show `task_id_pattern` and `dag_run_id_pattern` searches.

Same for dag_id

RoyLee1224 replied:
Sure, I'll implement it right away.

pierrejeambrun reviewed (commented):
A few things to address but looking good overall!

RoyLee1224 committed:
fix(api): keep xcom_key and map_index

RoyLee1224 committed:
fix(ui): improve XCom filter UX with NumberInput

RoyLee1224 committed:
fix(ui): showing only relevant filters in TI Xcom tab

RoyLee1224 committed:
fix(ui): replace unicode placeholder with CSS layout

RoyLee1224 committed:
Merge branch 'main' into filters

RoyLee1224 committed:
Merge branch 'main' into filters
